orx-mesh-generators/src/commonMain/kotlin/Extrusion.kt

```diff
@@ -0,0 +1,242 @@
+package org.openrndr.extra.meshgenerators
+
+import org.openrndr.math.Matrix44
+import org.openrndr.math.Vector2
+import org.openrndr.math.Vector3
+import org.openrndr.math.Vector4
+import org.openrndr.math.transforms.normalMatrix
+import org.openrndr.shape.Path3D
+import org.openrndr.shape.Shape
+import org.openrndr.shape.ShapeContour
+import org.openrndr.shape.Triangle
+
+fun quadToTris(v00: Vector3, v01: Vector3, v10: Vector3, v11: Vector3, faceNormal: Vector3, writer: VertexWriter) {
+
+    writer(v11, faceNormal, Vector2.ZERO)
+    writer(v01, faceNormal, Vector2.ZERO)
+    writer(v00, faceNormal, Vector2.ZERO)
+
+    writer(v00, faceNormal, Vector2.ZERO)
+    writer(v10, faceNormal, Vector2.ZERO)
+    writer(v11, faceNormal, Vector2.ZERO)
+}
+
+fun contourSegment(linearContour: List<Vector3>, frame0: Matrix44, frame1: Matrix44, writer: VertexWriter) {
+    for (i in linearContour.indices) {
+        val v0 = linearContour[i]
+        val v1 = linearContour[(i + 1).mod(linearContour.size)]
+
+        val v00 = (frame0 * v0.xyz1).xyz
+        val v01 = (frame0 * v1.xyz1).xyz
+
+        val v10 = (frame1 * v0.xyz1).xyz
+        val v11 = (frame1 * v1.xyz1).xyz
+        val faceNormal = ((v10 - v00).normalized cross (v01 - v00).normalized).normalized
+        quadToTris(v00, v01, v10, v11, faceNormal, writer)
+    }
+}
+
+fun triangulationWithFrame(triangulation: List<Triangle>, frame:Matrix44, invert:Boolean = true, writer:VertexWriter) {
+    val normalFrame = normalMatrix(frame)
+    val normalScale = if (!invert) -1.0 else 1.0
+    val normal = ((normalFrame * Vector4(0.0, 0.0, normalScale, 0.0)).xyz)
+    for (triangle in triangulation) {
+        val t = if (!invert) triangle else Triangle(triangle.x3, triangle.x2, triangle.x1)
+        writer((frame * t.x1.xy01).xyz, normal, Vector2.ZERO)
+        writer((frame * t.x2.xy01).xyz, normal, Vector2.ZERO)
+        writer((frame * t.x3.xy01).xyz, normal, Vector2.ZERO)
+    }
+}
+
+fun extrudeContourSteps(
+    contour: ShapeContour, path: Path3D, stepCount: Int, up0: Vector3,
+    contourDistanceTolerance: Double = 0.5,
+    pathDistanceTolerance: Double = 0.5,
+    steps: List<Vector3> = path.equidistantPositions(stepCount, pathDistanceTolerance),
+    frames: List<Matrix44> = steps.frames(up0),
+    startCap: Boolean = true,
+    endCap: Boolean = true,
+    writer: VertexWriter
+) {
+    val linearContour = contour.sampleLinear(contourDistanceTolerance)
+    val linearContourPoints = linearContour.adaptivePositions().map { it.xy0 }
+
+    extrudeCaps(linearContour.shape, path, startCap, endCap, frames, writer)
```

Comment:
What about not doing caps if the shape is closed?

Reply:
`extrudeCaps` checks for `path.closed`